app/models/contribution.rb

```diff
@@ -114,6 +140,33 @@ def age_and_consent_exclusivity
     errors.add(:age_confirm, I18n.t('contribute.campaigns.migration.form.validation.age_and_consent_exclusivity')) if age_confirm? && guardian_consent?
   end
 
+  # Derive an OAI-PMH resumption token for this contribution
+  #
+  # Concatenates ISO8601-formatted +first_published_at+, "/" and +oai_pmh_record_id+.
+  #
+  # @return [String]
+  def derive_oai_pmh_resumption_token
+    xml_time = first_published_at.iso8601
+    xml_time.sub!(/[+-]00:00\z/, 'Z') # this should be done by +Time#iso8601+ but seems not
+    xml_time + '/' + oai_pmh_record_id
+  end
+
+  # Set derived OAI-PMH fields
+  #
+  # It is ugly running this from an after_save callback, but it can not go into
+  # the AASM event because the CHO may not be persisted yet, hence not yet have
+  # a UUID.
+  def set_oai_pmh_fields
+    return if @setting_oai_pmh_fields
+    @setting_oai_pmh_fields = true
+
+    self.oai_pmh_record_id = ore_aggregation.edm_aggregatedCHO.uuid if self.oai_pmh_record_id.nil?
+    self.oai_pmh_resumption_token = derive_oai_pmh_resumption_token if self.oai_pmh_resumption_token.nil?
+    save
```

Comment:
potentially use skip_callback and then re-enable around this save instead of assigning the instance variable.

Reply:
although while the instance variable here seems a bit more complicated, I think it's probably safer in case something goes wrong during the save.

Reply:
How would you envisage that working given that `skip_callback` is a class method?

Reply:
something like:
```
    ...
    self.class.skip_callback(:save, :after, :set_oai_pmh_fields)
    save
    self.class.set_callback(:save, :after, :set_oai_pmh_fields)
```
but, like I said, this doesn't seem safe for when something goes wrong during the save, and handling that would make it as complex if not more than keeping track of the instance variable.